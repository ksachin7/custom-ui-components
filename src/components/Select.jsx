import React, { useState } from 'react'; // eslint-disable-line no-unused-vars

import PropTypes from 'prop-types';
import styled from 'styled-components';
import { MdArrowDropDown } from "react-icons/md";

// Styled components
const SelectContainer = styled.div`
  position: relative;
  display: inline-block;
`;

const SelectDropdown = styled.select`
  appearance: none;
  padding: 8px 32px 8px 12px; /* Adjust padding for the arrow button */
  font-size: 16px;
  border: 1px solid var(--color-grey-100);
  border-radius: 4px;
  background-color: var(--color-grey-100);
  cursor: pointer;
  width: ${props => props.width || 'auto'}; /* Customize width if needed */
  &:focus{
    /* border: none; */
    outline: none;
  }
`;

const ArrowIcon = styled.span`
  position: absolute;
  top: 56%;
  right: 10px;
  transform: translateY(-50%);
  pointer-events: none; /* Ensures the arrow icon doesn't interfere with select functionality */
`;

// Select component
const Select = ({ options, defaultValue, onChange, width }) => {
  const [selectedValue, setSelectedValue] = useState(defaultValue);
  const [isOptionsOpen, setIsOptionsOpen] = useState(false);

  const handleSelectChange = (event) => {
    const value = event.target.value;
    setSelectedValue(value);
    if (onChange) {
      onChange(value);
    }
  };

  const toggleOptions = () => {
    setIsOptionsOpen(!isOptionsOpen);
  };

  return (
    <SelectContainer>
      <SelectDropdown
        data-testid='Select-Dropdown'
        value={selectedValue}
        onChange={handleSelectChange}
        width={width}
        onClick={toggleOptions}
      >
        {options.map((option, index) => (
          <option key={index} value={option.value}>
            {option.label}
          </option>
        ))}
      </SelectDropdown>
      <ArrowIcon><MdArrowDropDown /></ArrowIcon>
    </SelectContainer>
  );
};

Select.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ).isRequired,
  defaultValue: PropTypes.string,
  onChange: PropTypes.func,
  width: PropTypes.string,
};

export default Select;
