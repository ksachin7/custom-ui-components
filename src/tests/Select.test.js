import React from 'react'; // eslint-disable-line no-unused-vars
import { shallow } from 'enzyme'; 
import Select from '../components/Select';

describe('Select component', () => {
  const options = [
    { label: 'Option 1', value: 'option1' },
    { label: 'Option 2', value: 'option2' },
    { label: 'Option 3', value: 'option3' },
  ];

  it('renders with default props', () => {
    const wrapper = shallow(<Select options={options} />); // Use shallow instead of mount
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with custom props', () => {
    const defaultValue = 'option2';
    const onChange = jest.fn();
    const wrapper = shallow(<Select options={options} defaultValue={defaultValue} onChange={onChange} width="200px" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('toggles dropdown when clicked', () => {
    const wrapper = shallow(<Select options={options} />);
    // console.log(wrapper.debug())
    // Find the styled select component
    const styledSelect = wrapper.find("[data-testid='Select-Dropdown']");

    // Simulate click to open dropdown
    styledSelect.simulate('click');

    // Assert the value of the selected option
    expect(styledSelect.find('option').at(0).prop('value')).toBe(options[0].value); // Dropdown should remain closed

    // Simulate click to close dropdown
    wrapper.find("[data-testid='Select-Dropdown']").simulate('click');
    // expect(wrapper.find('select').prop('value')).toBe(options[0].value); // Dropdown should remain closed
  });
});
