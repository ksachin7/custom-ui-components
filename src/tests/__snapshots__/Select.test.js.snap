// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component renders with custom props 1`] = `ShallowWrapper {}`;

exports[`Select component renders with default props 1`] = `ShallowWrapper {}`;
